contracts/strategies/BenqiStrategyV1.sol

```diff
@@ -0,0 +1,356 @@
+        require(tokenDelegator.redeemUnderlying(toWithdraw) == 0, "BenqiStrategyV1::redeem failed");
+        if (newBalance > 0) {
+            _rollupDebt(newBalance, 0);
+        }
+        
+        return toWithdraw;
+    }
+
+    function reinvest() external override onlyEOA {
+        uint unclaimedRewards = checkReward();
+        require(unclaimedRewards >= MIN_TOKENS_TO_REINVEST, "BenqiStrategyV1::reinvest");
+        _reinvest(unclaimedRewards);
+    }
+
+    receive() external payable {
+        require(msg.sender == address(rewardController), "BenqiStrategyV1::payments not allowed");
+    }
+
+    /**
+     * @notice Reinvest rewards from staking contract to deposit tokens
+     * @dev Reverts if the expected amount of tokens are not returned from `stakingContract`
+     * @param amount deposit tokens to reinvest
+     */
+    function _reinvest(uint amount) private {
+        rewardController.claimReward(0, address(this));
+        rewardController.claimReward(1, address(this));
+        WAVAX.deposit{value: address(this).balance}();
+        uint qiBalance = rewardToken0.balanceOf(address(this));
+        DexLibrary.swap(qiBalance, address(rewardToken0), address(rewardToken1), swapPairToken0);
+        amount = rewardToken1.balanceOf(address(this));
+
+        uint devFee = amount.mul(DEV_FEE_BIPS).div(BIPS_DIVISOR);
+        if (devFee > 0) {
+            _safeTransfer(address(rewardToken), devAddr, devFee);
+        }
+
+        uint adminFee = amount.mul(ADMIN_FEE_BIPS).div(BIPS_DIVISOR);
+        if (adminFee > 0) {
+            _safeTransfer(address(rewardToken), owner(), adminFee);
+        }
+
+        uint reinvestFee = amount.mul(REINVEST_REWARD_BIPS).div(BIPS_DIVISOR);
+        if (reinvestFee > 0) {
+            _safeTransfer(address(rewardToken), msg.sender, reinvestFee);
+        }
+        uint depositTokenAmount = DexLibrary.swap(
+            amount.sub(devFee).sub(adminFee).sub(reinvestFee),
+            address(rewardToken1),
+            address(depositToken),
+            swapPairToken1
+        );
+
+        _stakeDepositTokens(depositTokenAmount);
+        totalDeposits = totalDeposits.add(depositTokenAmount);
+
+        emit Reinvest(totalDeposits, totalSupply);
+    }
+
+    function _rollupDebt(uint principal, uint borrowed) internal {
+        (uint borrowLimit, uint borrowBips) = _getBorrowLimit();
+        uint supplied = principal;
+        uint lendTarget = principal.sub(borrowed).mul(leverageLevel).div(leverageBips);
+        uint totalBorrowed = borrowed;
+        while(supplied < lendTarget) {
+            uint toBorrowAmount = _getBorrowable(supplied, totalBorrowed, borrowLimit, borrowBips);
+            if (supplied.add(toBorrowAmount) > lendTarget) {
+                toBorrowAmount = lendTarget - supplied;
+            }
+            // safeguard needed because we can't mint below a certain threshold
+            if (toBorrowAmount < minMinting) {
+                toBorrowAmount = minMinting;
```

Comment:
would this *raise* the borrow amount, potentially to somewhere we don't want it? Maybe it's better to stop iterating if we are trying to borrow less than the minMinting instead of continue